File: examples/python/03_blend_comparison.py
"""Compare parabolic vs Hermite blend shapes on the same S-curve waypoints."""

import numpy as np

import justblend as jb


def run_one(shape: jb.BlendShape) -> None:
    waypoints = np.array(
        [
            [0.0, 0.0, 0.0],
            [1.0, 0.5, -0.3],
            [1.5, -0.2, 0.4],
            [0.5, 0.8, 1.0],
            [0.0, 0.0, 0.0],
        ]
    )
    limits = jb.Limits(
        v_max=np.array([1.5, 1.2, 1.0]),
        a_max=np.array([3.0, 2.5, 2.0]),
        j_max=np.array([20.0, 18.0, 15.0]),
    )
    opts = jb.GenerationOptions(
        blend_radius=0.15,
        corner_handling=jb.CornerHandling.HYBRID,
        blend_shape=shape,
    )

    gen = jb.SCurveTrajectoryGenerator(dim=3)
    gen.set_limits(limits)
    gen.set_options(opts)
    traj = gen.generate(waypoints)

    dense = traj.samples(dt=0.002)
    qdd_steps = np.max(np.abs(np.diff(dense.qdd, axis=0)))

    name = "parabolic" if shape == jb.BlendShape.PARABOLIC else "hermite"
    print(f"{name:<9}  duration={traj.duration():.4f} s  max |Δqdd| between adjacent dt=2ms = {qdd_steps:.4f}")


def main() -> None:
    run_one(jb.BlendShape.PARABOLIC)
    run_one(jb.BlendShape.HERMITE)


if __name__ == "__main__":
    main()

File: examples/cpp/03_blend_comparison.cpp
// 03_blend_comparison.cpp
//
// Run the same waypoints with both blend shapes and print summary statistics
// for the resulting trajectory. Hermite blends are slightly slower (smaller
// allowed corner speed because the peak in-blend acceleration is 1.5x
// parabolic for the same V and r) but have zero acceleration at the blend
// boundaries -- so paired with an S-curve linear profile there is no
// acceleration step at the linear/blend junctions.

#include <Eigen/Core>
#include <cstdio>

#include <justblend/justblend.hpp>

namespace
{

void runOne(justblend::BlendShape shape)
{
    Eigen::MatrixXd waypoints(5, 3);
    waypoints << 0.0, 0.0, 0.0, 1.0, 0.5, -0.3, 1.5, -0.2, 0.4, 0.5, 0.8, 1.0, 0.0, 0.0, 0.0;

    justblend::Limits limits;
    limits.v_max = (Eigen::VectorXd(3) << 1.5, 1.2, 1.0).finished();
    limits.a_max = (Eigen::VectorXd(3) << 3.0, 2.5, 2.0).finished();
    limits.j_max = (Eigen::VectorXd(3) << 20.0, 18.0, 15.0).finished();

    justblend::GenerationOptions opts;
    opts.blend_radius = 0.15;
    opts.corner_handling = justblend::CornerHandling::Hybrid;
    opts.blend_shape = shape;

    justblend::SCurveTrajectoryGenerator gen(3);
    gen.setLimits(limits);
    gen.setOptions(opts);
    auto traj = gen.generate(waypoints);

    auto dense = traj.samples(0.002);
    double max_step = 0.0;
    for (Eigen::Index i = 1; i < dense.qdd.rows(); ++i)
    {
        max_step = std::max(max_step, (dense.qdd.row(i) - dense.qdd.row(i - 1)).cwiseAbs().maxCoeff());
    }
    const char* name = shape == justblend::BlendShape::Parabolic ? "parabolic" : "hermite";
    std::printf("%-9s  duration=%.4f s  max |Δqdd| between adjacent dt=2ms = %.4f\n", name, traj.duration(), max_step);
}

} // namespace

int main()
{
    runOne(justblend::BlendShape::Parabolic);
    runOne(justblend::BlendShape::Hermite);
    return 0;
}
